troi/listenbrainz/dataset_fetcher.py

```diff
@@ -13,62 +13,99 @@ class DataSetFetcherElement(Element):
+    def __init__(self, server_url, json_post_data, max_num_items=None):
         super().__init__()
         self.server_url = server_url
         self.json_post_data = json_post_data
+        self.max_num_items = max_num_items
 
     @staticmethod
     def inputs():
         return []
 
     @staticmethod
     def outputs():
-        return [Recording]
+        return [Recording, Artist]
 
-    def read(self, inputs):
+    def create_recording(self, row):
 
-        r = requests.post(self.server_url, json=self.json_post_data)
-        if r.status_code != 200:
-            raise PipelineError("Cannot fetch first dataset recordings from ListenBrainz. HTTP code %s (%s)" % (r.status_code, r.text))
+        r = Recording(mbid=row['recording_mbid'])
+        if 'artist_credit_name' in row:
+            if 'artist_credit_id' in row:
+                r.artist.artist_credit_id = row['artist_credit_id']
 
-        recordings = []
-        for row in r.json():
-            if row['recording_mbid'] is None:
-                continue
+        artist = None
+        if 'artist_credit_name' in row:
+            artist = Artist(name=row['artist_credit_name'])
 
-            r = Recording(mbid=row['recording_mbid']) 
-            if 'artist_credit_name' in row:
-                if 'artist_credit_id' in row:
-                    r.artist.artist_credit_id = row['artist_credit_id']
+        if 'artist_credit_id' in row:
+            if artist is None:
+                artist = Artist(artist_credit_id=row['artist_credit_id'])
+            else:
+                artist.artist_credit_id = row['artist_credit_id']
 
-            artist = None
-            if 'artist_credit_name' in row:
-                artist = Artist(name=row['artist_credit_name'])
+        if 'artist_mbids' in row:
+            if artist is None:
+                artist = Artist(mbids=row['artist_mbids'])
+            else:
+                artist.mbids = row['artist_mbids']
 
-            if 'artist_credit_id' in row:
-                if artist is None:
-                    artist = Artist(artist_credit_id=row['artist_credit_id'])
-                else:
-                    artist.artist_credit_id = row['artist_credit_id']
+        r.artist = artist
+
+        if 'recording_name' in row:
+            r.name = row['recording_name']
+
+        if 'year' in row:
+            r.year = row['year']
+
+        if 'listen_count' in row:
+            r.listenbrainz = {"listen_count": row["listen_count"]}
+
+        if 'bpm' in row:
+            r.acousticbrainz = {"bpm": row["bpm"]}
 
-            if 'artist_mbids' in row:
-                if artist is None:
-                    artist = Artist(mbids=row['artist_mbids'])
-                else:
-                    artist.mbids = row['artist_mbids']
+        return r
 
-            r.artist = artist
+    def create_artist(self, row):
 
-            if 'recording_name' in row:
-                r.name = row['recording_name']
+        if 'artist_mbid' in row:
```

Comment:
Instead of returning Artist directly, maybe create an Artist object and try to set all items in it. Also, reuse this method in create_recording?

Reply:
I dont understand what you're saying in the first sentence. Please elaborate.

Reply:
I mean do it the the same way as in `create_recording`, `a = Artist()` then each if checks and sets the relevant field of the artist object.

Reply:
I tried that, but the code ends up being more complicated that way, since there is a possibility of returning None, rather than an empty Artist object. I prefer to keep as is.